src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -209,9 +238,72 @@ ProjNode* PhaseIdealLoop::create_new_if_for_predicate(ProjNode* cont_proj, Node*
   return if_cont->as_Proj();
 }
 
+// Clone data nodes for the fast loop while creating a new If with create_new_if_for_predicate. Returns the node which is
+// used for the uncommon trap phi input.
+Node* PhaseIdealLoop::clone_data_nodes_for_fast_loop(Node* phi_input, ProjNode* uncommon_proj, Node* if_uct, Node_List* old_new) {
+  // Step 1: Clone all nodes on the data chain but do not rewire anything, yet. Keep track of the cloned nodes
+  // by using the old_new mapping. This mapping is then used in step 2 to rewire the cloned nodes accordingly.
+  uint last_idx = C->unique();
+  Unique_Node_List list;
+  list.push(phi_input);
+  for (uint j = 0; j < list.size(); j++) {
+    Node* next = list.at(j);
+    Node* clone = next->clone();
+    _igvn.register_new_node_with_optimizer(clone);
+    old_new->map(next->_idx, clone);
+    if (next->in(0) == uncommon_proj) {
```

Comment:
Doesn't that assume only one node in the chain has its control set? Wouldn't testing something like get_ctrl(next) != uncommon_proj be more robust?

Reply:
Nice catch! This assumption is indeed wrong. I played around with some `Div` nodes and the algorithm bails out in step 2 too early when comparing for `next->in(0) == uncommon_proj` without doing a rewiring. This leads to a bad graph again. 

I think this check here in step 1 can be completely removed due to already checking with `get_ctrl()` below. Step 2 needs to be adapted such that we rewire the control input of a data nodes but also apply the rewiring process for such nodes with a control input within a chain. I pushed an update.